csrc/device_lower/analysis/tensor_memory.cpp

```diff
@@ -13,27 +13,82 @@
 
 namespace nvfuser {
 
+// See note [Tensor Memory Allocation] for the overall design.
 TensorMemoryInfo computeTMemInfo(Fusion* fusion) {
-  bool found = false;
+  TensorMemoryInfo result;
+
+  // Step 1: partition the tensors. Each partition of tensor will become a
+  // region, so we use the term partition and region interchangeably. The user
+  // may have provided full or partial partitioning information. For the
+  // TensorViews that the user has already specified which region they belong
+  // to, we will use that information. For the rest of the tensors, we will
+  // assign each of them to a separate region.
+  using Partition = std::vector<std::vector<TensorView*>>;
+  Partition partitions;
+  if (fusion->hasManaged("tmem_regions")) {
+    partitions = fusion->getManaged<Partition>("tmem_regions");
+  } else {
+    partitions = {};
+  }
+
+  // Verify that there is no overlap between user specified partitions
+  std::unordered_set<TensorView*> tensors;
+  for (auto& partition : partitions) {
+    NVF_ERROR(!partition.empty(), "Empty partition");
+    for (auto tv : partition) {
+      NVF_ERROR(
+          tv->getMemoryType() == MemoryType::Tensor, "Invalid memory type");
+      NVF_ERROR(
+          tensors.insert(tv).second, "Tensors cannot be in multiple regions");
+    }
+  }
+
+  // For all TensorViews whose partition is not specified, assign them to a
+  // separate region.
   for (auto tv : fusion->allTvs()) {
-    if (tv->getMemoryType() == MemoryType::Tensor) {
-      NVF_ERROR(!found, "Only one tensor on TMem is supported");
-      found = true;
+    if (tv->getMemoryType() != MemoryType::Tensor) {
+      continue;
+    }
+    if (tensors.count(tv) == 0) {
+      partitions.push_back({tv});
     }
   }
 
-  if (found) {
+  // Step 2: Compute the allocation information for tensor memory. That is, for
+  // each partition, we create a Region object and fill in the necessary
+  // information.
+  auto& regions = result.allocation.regions;
```

Comment:
nit. I believe this would help read the code as I didn't remember the type and needed go back and forth between this and the header file.

```suggestion
  std::vector<Region>& regions = result.allocation.regions;
```

Reply:
changed